hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/SchemaReplicator.java

```diff
@@ -0,0 +1,364 @@
+     * {@code replications}.
+     * <p>
+     * It assumes that the {@code this.replications} is an empty map.
+     * <p>
+     * This method might be called on two occasions:
+     * <ul>
+     *     <li>For newly joining members, where {@code this.replications} is
+     *     empty</li>
+     *     <li>For members of the smaller cluster when they join the larger
+     *     cluster, during the split brain healing, where
+     *     {@link #clear()} is called beforehand, which clears the
+     *     {@code this.replications}.</li>
+     * </ul>
+     */
+    public void setReplications(Collection<SchemaReplication> replications) {
+        for (SchemaReplication replication : replications) {
+            long schemaId = replication.getSchema().getSchemaId();
+            this.replications.put(schemaId, replication);
+        }
+    }
+
+    private void doReplicate(Schema schema, CompletableFuture<Void> future) {
+        long schemaId = schema.getSchemaId();
+        try {
+            prepareOnCaller(schema)
+                    .thenCompose(result -> {
+                        markSchemaAsPrepared(schema);
+                        return sendRequestForPreparation(schema);
+                    })
+                    .thenCompose(result -> sendRequestForAcknowledgment(schemaId))
+                    .thenRun(() -> completeInFlightOperation(schemaId, future))
+                    .exceptionally(throwable -> {
+                        completeInFlightOperationExceptionally(schemaId, future, throwable);
+                        return null;
+                    });
+        } catch (Throwable t) {
+            // to avoid risk of prepareOnCaller throwing synchronously
+            completeInFlightOperationExceptionally(schemaId, future, t);
+        }
+    }
+
+    private CompletableFuture<Void> prepareOnCaller(Schema schema) {
+        schemaService.putLocal(schema);
+        return schemaService.persistSchemaToHotRestartAsync(schema);
+    }
+
+    private void doReplicatePreparedSchema(Schema schema, CompletableFuture<Void> future) {
+        long schemaId = schema.getSchemaId();
+        try {
+            sendRequestForPreparation(schema)
+                    .thenCompose(result -> sendRequestForAcknowledgment(schemaId))
+                    .thenRun(() -> completeInFlightOperation(schemaId, future))
+                    .exceptionally(throwable -> {
+                        completeInFlightOperationExceptionally(schemaId, future, throwable);
+                        return null;
+                    });
+        } catch (Throwable t) {
+            // to avoid risk of sendRequestForPreparation throwing synchronously
+            completeInFlightOperationExceptionally(schemaId, future, t);
+        }
+    }
+
+    private void completeInFlightOperation(long schemaId, CompletableFuture<Void> future) {
+        synchronized (mutex) {
+            markSchemaAsReplicated(schemaId, false);
```

Comment:
No, it is the other way around. If `markSchemaAsReplicated` is called from the `AckSchemaReplicationOperation`, this will be true. For others, this will be `false`. We don't want to mark the local schema as REPLICATED in the local execution of `AckSchemaReplicationOperation`. (invokeOnStableClusterSerial executes the operations locally as well)

Reply:
may be the parameter name `isLocalOperationExecution ` is misleading.